src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala

```diff
@@ -69,7 +76,25 @@ object Reader {
 
     System.setProperty(LineReader.PROP_SUPPORT_PARSEDLINE, java.lang.Boolean.TRUE.toString())
 
-    val jlineTerminal = TerminalBuilder.builder().jna(true).build()
+    val jlineTerminal = {
+      /*
+       * sbt needs to override the default jline terminal both to prevent the creation
+       * of two jline3 system terminals, which interferses with sbt's jline3 reader, and
+       * to handle the case where a remote client is running the console and sbt needs to
+       * relay io between the client and the server process. More broadly, this allows the
+       * repl to be embedded in any application that is also using jline3.
+       */
+      System.getProperty("scala.jline.terminal.provider") match {
+        case null => TerminalBuilder.builder().jna(true).build()
+        case c =>
+          try {
+            val clazz = Thread.currentThread.getContextClassLoader.loadClass(c)
+            val method = clazz.getDeclaredMethod("getJLine3Terminal")
+            val instance = clazz.getDeclaredField("MODULE$").get(null)
+            method.invoke(instance).asInstanceOf[Terminal]
```

Comment:
This looks like it requires sbt and the compiler to use binary-compatible versions of JLine, I don't think that's acceptable since it effectively means that neither sbt nor the compiler can ever upgrade JLine: sbt needs to work with multiple versions of the compiler, and the compiler needs to work with multiple versions of sbt.

Reply:
Also I glanced at the config and I couldn't remember if setting up the Reader changes state of terminal. One can see unfortunate interactions (which also plague current sbt/console interactions). OTOH, sbt is always more worried that it must work with all versions of scala, which I never understood as an ultimatum.

Reply:
Also, there's already a mechanism for sbt to communicate with the compiler in a binary-compatible way: the compiler-bridge. It already needs to deal with the REPL in some ways: https://github.com/sbt/zinc/blob/develop/internal/compiler-bridge/src/main/scala_2.13/xsbt/ConsoleInterface.scala, I don't fully understand your usecase but maybe it could be extended to support it?

Reply:
This "required by sbt" is dearest to my heart: https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/Reporting.scala#L236

I think REPL front end is more likely to have a stable API in its Reader abstraction, or InteractiveReader. But I don't remember without looking how completion is hooked up. There might be multiple opportunities for tight coupling.

Reply:
> This looks like it requires sbt and the compiler to use binary-compatible versions of JLine, I don't think that's acceptable since it effectively means that neither sbt nor the compiler can ever upgrade JLine: sbt needs to work with multiple versions of the compiler, and the compiler needs to work with multiple versions of sbt.

This is a theoretical problem with or without this change because sbt upgraded to JLine3. As a result, the sbt version of JLine will be used in the embedded scala console no matter what. This is the cost of embedding the console in the sbt server process. The only way to avoid the conflict would be to require forking the `console` task for scala 2.13.x and presumably scala 3. sbt doesn't presently support forking the console task and it would be a pain to implement.

sbt needs to do some very specific things to make the thin client work that are difficult to embed within the compiler bridge without actually exposing the jline Terminal in the api or just creating a mirror interface that we have to convert to and from between sbt and the scala repl. JLine needs to know properties of the terminal such as the height and width to work correctly. The dimensions of the client terminal may not be the same as the terminal running the server (if it is even being run in a tty). As a result, sbt implements a protocol that allows it to query the remote client for the terminal properties (there are many besides height and width).

In addition to the thin client, JLine is a very finicky library and I had to put a lot of effort into making sure that it didn't mess up the terminal state. In sbt we also have to concurrently support JLine 2 in order for the console to work with scala 2.12.x. Entering the LineReader using the default modifies the global properties of the underlying tty emulator which, as I mentioned, can cause problems for sbt. In order for sbt to interoperate with the scala REPL, it is necessary for some coordination to exist to avoid these problems. It is very unfortunate that the two are intertwined in this way, but there is no good way to avoid that.

As a result, it will always be the case that if scala decides to upgrade its version of JLine 3 and use backward incompatible methods then it won't work when embedded in sbt until a new version of sbt is published with the new version of JLine. I find it somewhat unlikely that that scala will ever use such a binary incompatible method. If it ever did, I don't think it's unreasonable to force users to upgrade sbt to use the console for the new scala version. JLine 3 preserves backward compatibility so it is extremely unlikely that an sbt upgrade would break the scala console (from a binary compatibility standpoint at least).

It would be possible to pass a new abstract `Terminal` interface into the `ConsoleInterface` and `ILoop` but that would require more or less making a duplicate of https://github.com/sbt/sbt/blob/e985d77ace9f4950a9586c3cd9fb20fd23ef1039/internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala#L22 inside of ILoop and adding a parameter to `ConsoleInterface.run`. There would be a ton of boilerplate across multiple repos that, as a volunteer, I am not personally excited about implementing myself. It would likely total hundreds of lines of code across multiple repositories.

What is good about this change is that it is small, scoped to a single file and only exercised in the presence of a system property. The blast radius is very small. Sure, it is a bit of a hack but it will not impact any other downstream consumers of the scala repl. It also is easy to disable on the sbt side by not setting the sytem property. In my opinion, the benefit is very high because it allows people to seamlessly enter a scala console from their shell without having to start sbt (provided they have started an sbt server process once).

Reply:
> This is a theoretical problem with or without this change because sbt upgraded to JLine3. As a result, the sbt version of JLine will be used in the embedded scala console no matter what. This is the cost of embedding the console in the sbt server process. The only way to avoid the conflict would be to require forking the console task for scala 2.13.x and presumably scala 3. sbt doesn't presently support forking the console task and it would be a pain to implement.

So sbt is mixing its own classpath with the compiler classpath? That seems like a problem indeed, but instead of forking, can't this be solved using classloaders?

Reply:
> So sbt is mixing its own classpath with the compiler classpath? That seems like a problem indeed, but instead of forking, can't this be solved using classloaders?

Yes. But we explicitly don't want that because having two instance of the jline system terminal can cause problems with process because they both modify the global state of the program terminal.

Reply:
This is a bit confusing to talk about because there is the jline `Terminal` class and the underlying terminal emulator. Regardless of whether or not sbt and scala repl have different versions of the library, they both can modify the underlying terminal emulator which is a problem. Part of the motivation for this change is to prevent the repl from doing that. The other part is to make the thin client work so that users can enter the scala console from the shell without sbt start up time.

Reply:
Speaking of the compiler bridge, `ConsoleInterface#run(...)` already takes a parameter called `loader: ClassLoader`,  the bridge implementation currently construct `IMain` with an overridden `parentClassLoader` -  https://github.com/sbt/zinc/blob/8c1e6661123177b25ce362cb5b1b0701f414f7fc/internal/compiler-bridge/src/main/scala_2.13/xsbt/ConsoleInterface.scala#L28-L38

If I understand correctly, between Zinc and Scala 2.13.3/3.x compiler I don't think any further communication is needed. What we need instead an entry point. For example, `ILoop#defaultIn`:

https://github.com/scala/scala/blob/a6e1a6f476f26f66ae1c3848033e95f6b173ad56/src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala#L63-L70

currently that's private, but if Scala could make that protected, then Ethan could instantiate his own `Reader` instance that behaves according to `InteractiveReader` contract. It might do some ClassLoader magic internally but that would be Zinc's concern.

Reply:
@eed3si9n The problem that I see with going through the compiler bridge is that it is defined in zinc, not sbt itself. There isn't an obviously great way to get an sbt generated terminal in it without making the compiler-bridge depend on jline 3 which isn't ideal. I am starting to think that the best approach is actually to fork jline3 the same way that we forked jline 2. I already got that working on my computer and that doesn't require any compiler bridge or scala changes and would automatically work with 2.13.x as well as dotty.

The drawback to the forked jline is that we have to maintain it and scala can't use newer jline 3 features from sbt without updating the forked sbt jline but that is going to be the case no matter how this functionality is implemented. The forked code is like 10 lines that are unlikely to have conflicts with jline3 so updating the fork would be pretty easy.

Reply:
> having two instance of the jline system Terminal can cause problems with process because they both modify the global state of the terminal running the program

What's the "jline system Terminal" and what can cause it to modify the global state?

I'm wondering whether it would be possible layer a newer JLine 3 when launching `console` if the Scala version brings a newer version in.

> The forked code is about 10 lines that are unlikely to have conflicts with jline3 so updating the fork would be pretty easy.

What's the change?

Reply:
```patch
diff --git a/reader/src/main/java/org/jline/keymap/BindingReader.java b/reader/src/main/java/org/jline/keymap/BindingReader.java
index 1c9a175d..e4553a4c 100644
--- a/reader/src/main/java/org/jline/keymap/BindingReader.java
+++ b/reader/src/main/java/org/jline/keymap/BindingReader.java
@@ -10,6 +10,7 @@ package org.jline.keymap;
 
 import java.io.IOError;
 import java.io.IOException;
+import java.io.EOFException;
 import java.util.ArrayDeque;
 import java.util.Deque;
 
@@ -166,6 +167,8 @@ public class BindingReader {
             return s != 0 ? Character.toCodePoint((char) s, (char) c) : c;
         } catch (ClosedException e) {
             throw new EndOfFileException(e);
+       } catch (EOFException e) {
+           throw new EndOfFileException(e);
         } catch (IOException e) {
             throw new IOError(e);
         }
diff --git a/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java b/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java
index a5b7c0a9..1ad50a41 100644
--- a/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java
+++ b/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java
@@ -260,6 +260,11 @@ public final class TerminalBuilder {
     }
 
     public Terminal build() throws IOException {
+       if (terminalHolder.get() != null) {
+         Terminal terminal = terminalHolder.get();
+         Log.debug(() -> "Using terminal " + terminal.getClass().getSimpleName());
+         return terminal;
+       }
         Terminal terminal = doBuild();
         Log.debug(() -> "Using terminal " + terminal.getClass().getSimpleName());
         if (terminal instanceof AbstractPosixTerminal) {
@@ -486,4 +491,9 @@ public final class TerminalBuilder {
     private <S> S load(Class<S> clazz) {
         return ServiceLoader.load(clazz, clazz.getClassLoader()).iterator().next();
     }
+
+    private static final AtomicReference<Terminal> terminalHolder = new AtomicReference<Terminal>(null);
+    public static final void setTerminal(final Terminal terminal) {
+      terminalHolder.set(terminal);
+    }
 }
```

Reply:
https://github.com/jline/jline3/pull/554

Reply:
> What's the "jline system Terminal" and what can cause it to modify the global state?

https://github.com/jline/jline3/blob/7e807af9bd7cb86228fbceb215488218faeca9df/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java#L63-L89
When you don't provide input and output streams to the builder, it creates a "system" terminal. When you enter a LineReader, JLine will issue a number of TTY commands through the system terminal that do things like change the input mode from canonical (System.in is buffered so that read doesn't return until a new line is entered and echo is enabled) to raw (System.in is not buffered, echo is disabled). I am not sure of all of the things it does but I have found that jline 3 modifying the TTY can break the scala console for 2.12.x (https://github.com/sbt/sbt/pull/5690).